tests/http/server/_base.py

```diff
@@ -98,3 +111,413 @@ def _test_disconnect(
+                else:
+                    # This `Deferred` does not have a result yet.
+                    # The `await` will block normally, so we don't have to do
+                    # anything.
+                    self._log_await_stack(
+                        stack, previous_stack, request_number, "blocking await"
+                    )
+                    previous_stack = stack
+
+                    return original_Deferred___next__(deferred, value)
+
+        return Deferred___next__, unblock_awaits, get_awaits_seen, has_seen_new_await
+
+    def _log_for_request(self, request_number: int, message: str) -> None:
+        """Logs a message for an iteration of `_test_cancellation_at_every_await`."""
+        # We want consistent alignment when logging stack traces, so ensure the
+        # logging context has a fixed width name.
+        with LoggingContext(name=f"request-{request_number:<2}"):
+            logger.info(message)
+
+    def _log_await_stack(
+        self,
+        stack: List[inspect.FrameInfo],
+        previous_stack: List[inspect.FrameInfo],
+        request_number: int,
+        note: str,
+    ) -> None:
+        """Logs the stack for an `await` in `_test_cancellation_at_every_await`.
+
+        Only logs the part of the stack that has changed since the previous call.
+
+        Example output looks like:
+        ```
+        delay_cancellation:750 (synapse/util/async_helpers.py:750)
+            DatabasePool._runInteraction:768 (synapse/storage/database.py:768)
+                > *blocked on await* at DatabasePool.runWithConnection:891 (synapse/storage/database.py:891)
+        ```
+
+        Args:
+            stack: The stack to log, as returned by `_get_stack()`.
+            previous_stack: The previous stack logged, with callers appearing before
+                callees.
+            request_number: The request number to log against.
+            note: A note to attach to the last stack frame, eg. "blocked on await".
+        """
+        for i, frame_info in enumerate(stack[:-1]):
+            # Skip any frames in common with the previous logging.
+            if i < len(previous_stack) and frame_info == previous_stack[i]:
+                continue
+
+            frame = self._format_stack_frame(frame_info)
+            message = f"{'  ' * i}{frame}"
+            self._log_for_request(request_number, message)
+
+        # Always print the final frame with the `await`.
+        # If the frame with the `await` started another coroutine run, we may have
+        # already printed a deeper stack which includes our final frame. We want to
+        # log where all `await`s happen, so we reprint the frame in this case.
+        i = len(stack) - 1
+        frame_info = stack[i]
+        frame = self._format_stack_frame(frame_info)
+        message = f"{'  ' * i}> *{note}* at {frame}"
+        self._log_for_request(request_number, message)
+
+    def _format_stack_frame(self, frame_info: inspect.FrameInfo) -> str:
```

Comment:
I think this and the other methods below can be turned into regular functions. There's no need for it all to be part of the same object.